### PR TITLE
Dev david issue 86

### LINKED ISSUE #86: teststrategy: infer lowest skill gap
The test ends if the lowest subscale (the one that underperforms significantly or to the greatest extent) can be determined via DPF
> „Der CAT-Test abgebrochen, wenn mittels DPF die unterste Subskala hinreichend bestimmt ist, die signifikant bzw. am stärksten unterperformt.“

scale to be tested selected via
- [x] PP is only calculated for scales that have a sufficient number of answers 
- [x] Abweichung ist PP (zu prüfende Skala) - PP (übergeordnete Skala)

additional abort conditions:
- standard error falls below a threshold
- maximum number of items to be viewed per scale is exceeded

### Details

Note:  A question can belong to both a scale and to one of its subscales
- Everytime a user gives an answer, calculate the person ability for the most specific subscale. At least two answers with differenct outcome (true/false) are needed in order to calculate the ability parameter.
- Selecting the next question works as follows:
    - If there is no information about ability per scale, just pick a question from the top-most scale via weighted fisher information. Otherwise:
    - Select the (sub)scale where the user has the lowest ability
   - From that scale, use the timepenalty-weighted fisher information to select the next question to be played

### PR DESCRIPTION
**Pull request by davidszkiba**
Closes #86

### CODE REVIEW

File: classes/output/attemptfeedback.php [hunk 1]
```diff
@@ -133,8 +133,17 @@ private function render_person_abilities() {
         $catscales = catquiz::get_catscales(array_keys($personabilities));
         $result = [];
         foreach ($personabilities as $catscaleid => $ability) {
+            if (abs(floatval($ability)) === abs(floatval(PERSONABILITY_MAX))) {
+                if ($ability < 0) {
+                    $ability = get_string('allquestionsincorrect', 'local_catquiz');
+                } else {
```
Comment (eynimeni) [inline classes/output/attemptfeedback.php:139]: why two cases?

File: classes/teststrategy/preselect_task/filterforlowestsubscale.php [hunk 2]
```diff
@@ -48,21 +49,28 @@ final class filterforlowestsubscale extends preselect_task implements wb_middlew
      *
      */
     public function run(array $context, callable $next): result {
-        // If there is no information about ability per scale, just pick a
-        // question from the top-most scale via weighted fisher information
         $abilities = $context['person_ability'];
-        asort($abilities);
-        $catscaleids = array_keys($abilities);
+        // The difference to itself is 0.
+        $abilitydifference = [$context['catscaleid'] => 0];
```
Comment (eynimeni) [inline classes/teststrategy/preselect_task/filterforlowestsubscale.php:54]: hard coded catscaleid?

File: classes/output/attemptfeedback.php (hunk 1 above)
Comment (davidszkiba) [reply classes/output/attemptfeedback.php:139]: They are different. One is for correct answers, one for incorrect answers.

File: classes/teststrategy/preselect_task/filterforlowestsubscale.php (hunk 2 above)
Comment (davidszkiba) [reply classes/teststrategy/preselect_task/filterforlowestsubscale.php:54]: Yes. Here I hardcode the ability difference to itself as 0.

**Pull request merged**